File: priv/static/blog_posts/algo/datastructures/stacks/implementation.go
package main
import "fmt"

type Stack struct {
  data []interface{}
  size int
}

func (s *Stack) push(item interface{}) {
  s.data = append(s.data, item)
  s.size += 1
}

func (s *Stack) pop() interface{} {
  if s.size == 0 {
    return nil
  }

  item := s.data[s.size-1]
  s.data = s.data[:s.size-1]
  s.size -= 1
  return item
}

func (s *Stack) peek() interface{} {
  return s.data[s.size-1]
}

func main() {
  s := Stack{}
  s.push(1)
  s.push(2)
  s.push(3)

  fmt.Println(s.data, "|", s.size)

  item := s.pop()
  fmt.Println(item)
  fmt.Println(s.data, "|", s.size)
  item = s.pop()
  fmt.Println(item)
  fmt.Println(s.data, "|", s.size)
}

File: priv/static/blog_posts/algo/datastructures/stacks/implementation.ex


defmodule Stack do
  @moduledoc """
  A simple stack implementation using Lists
  """

  defstruct [data: [], size: 0]

  def new do; %Stack{} end

  def push(stack, item) do
    # Because Elixir's lists are linked lists under the hood, we prepend rather than append.
    %Stack{
      data: [item | stack.data],
      size: stack.size + 1
    }
  end

  def pop(%Stack{ data: [] }) do; {:empty, new()} end
  def pop(%Stack{ data: [head] }) do; {head, new()} end
  def pop(%Stack{ data: [head | tail], size: size }) do
    {head, %Stack{data: tail, size: size-1}}
  end

  def peek(%Stack{}) do; :empty end
  def peek(%Stack{data: [head | _tail]}) do; head end
end
